upload/extension/opencart/catalog/model/payment/opayo.php

```diff
@@ -0,0 +1,707 @@
+		$query = $this->db->query("SELECT * FROM `" . DB_PREFIX . "opayo_order_subscription` WHERE `subscription_id` = '" . (int)$subscription_id . "'");
+
+		return $query->row;
+	}
+
+	/**
+	 * Add Subscription Transaction
+	 *
+	 * @param int   $subscription_id
+	 * @param int   $order_id
+	 * @param array $response_data
+	 * @param int   $type
+	 *
+	 * @return void
+	 */
+	private function addSubscriptionTransaction(int $subscription_id, int $order_id, array $response_data, int $type): void {
+		$this->db->query("INSERT INTO `" . DB_PREFIX . "subscription_transaction` SET `subscription_id` = '" . (int)$subscription_id . "', `order_id` = '" . (int)$order_id . "', `date_added` = NOW(), `amount` = '" . (float)$response_data['Amount'] . "', `type` = '" . (int)$type . "', `reference` = '" . $this->db->escape($response_data['VendorTxCode']) . "'");
+	}
+
+	/**
+	 * Get Profiles
+	 *
+	 * @return array
+	 */
+	private function getProfiles(): array {
+		$query = $this->db->query("SELECT `or`.`order_recurring_id` FROM `" . DB_PREFIX . "order_recurring` `or` JOIN `" . DB_PREFIX . "order` `o` USING(`order_id`) WHERE `o`.`payment_code` = 'opayo'");
+
+		$subscriptions = [];
+
+		foreach ($query->rows as $subscription) {
+			$subscriptions[] = $this->getProfile($subscription['subscription_id']);
+		}
+
+		return $subscriptions;
+	}
+
+	/**
+	 * Get Profile
+	 *
+	 * @param int $subscription_id
+	 *
+	 * @return array
+	 */
+	private function getProfile(int $subscription_id): array {
+		$query = $this->db->query("SELECT * FROM `" . DB_PREFIX . "subscription` WHERE `subscription_id` = " . (int)$subscription_id);
+
+		return $query->row;
+	}
+
+	/**
+	 * Update Cron Run Time
+	 *
+	 * @return void
+	 */
+	public function updateCronRunTime(): void {
+		$this->db->query("DELETE FROM `" . DB_PREFIX . "setting` WHERE `code` = 'opayo' AND `key` = 'payment_opayo_last_cron_run'");
+		$this->db->query("INSERT INTO `" . DB_PREFIX . "setting` (`store_id`, `code`, `key`, `value`, `serialized`) VALUES (0, 'opayo', 'payment_opayo_last_cron_run', NOW(), 0)");
+	}
+
+	/**
+	 * Send Curl
+	 *
+	 * @param string $url
+	 * @param array  $payment_data
+	 * @param ?null  $i
+	 *
+	 * @return array
+	 */
+	public function sendCurl(string $url, array $payment_data, $i = null): array {
+		$curl = curl_init($url);
+
+		curl_setopt($curl, CURLOPT_PORT, 443);
+		curl_setopt($curl, CURLOPT_HEADER, 0);
+		curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, 0);
+		curl_setopt($curl, CURLOPT_RETURNTRANSFER, 1);
+		curl_setopt($curl, CURLOPT_FOLLOWLOCATION, false);
+		curl_setopt($curl, CURLOPT_FORBID_REUSE, 1);
+		curl_setopt($curl, CURLOPT_FRESH_CONNECT, 1);
+		curl_setopt($curl, CURLOPT_POST, 1);
+		curl_setopt($curl, CURLOPT_POSTFIELDS, http_build_query($payment_data));
+
+		$response = curl_exec($curl);
+
+		curl_close($curl);
+
+		$data = [];
+
+		$response_info = explode(chr(10), $response);
+
+		foreach ($response_info as $string) {
+			if (strpos($string, '=') && isset($i)) {
```

Comment:
`$i` is defined in the signature so i always set.
```suggestion
			if (strpos($string, '=') && $i !== null) {
```